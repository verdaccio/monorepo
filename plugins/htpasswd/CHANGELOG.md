# Change Log

## 10.4.0

### Minor Changes

- 49ca26d: feat: allow other password hashing algorithms

  copied from v6 plugins by @greshilov https://github.com/verdaccio/verdaccio/pull/2072

  > To avoid a breaking change, the default algorithm is `crypt`.

  ### Context

  The current implementation of the `htpasswd` module supports multiple hash formats on verify, but only `crypt` on sign in.
  `crypt` is an insecure old format, so to improve the security of the new `verdaccio` release we introduce the support of multiple hash algorithms on sign in step.

  ### New hashing algorithms

  The new possible hash algorithms to use are `bcrypt`, `md5`, `sha1`. You can read more about them [here](https://httpd.apache.org/docs/2.4/misc/password_encryptions.html).

  Two new properties are added to `auth` section in the configuration file:

  - `algorithm` to choose the way you want to hash passwords.
  - `rounds` is used to determine `bcrypt` complexity. So one can improve security according to increasing computational power.

  Example of the new `auth` config file section:

  ```yaml
  auth:
  htpasswd:
    file: ./htpasswd
    max_users: 1000
    # Hash algorithm, possible options are: "bcrypt", "md5", "sha1", "crypt".
    algorithm: bcrypt
    # Rounds number for "bcrypt", will be ignored for other algorithms.
    rounds: 10
  ```

## 10.3.1

### Patch Changes

- Updated dependencies [b5cfaf6]
  - @verdaccio/file-locking@10.3.0

## 10.3.0

### Minor Changes

- 56c65d9: Refactor htpasswd plugin to use the bcryptjs 'compare' api call instead of 'comparSync'. Add a new configuration value named 'slow_verify_ms' to the htpasswd plugin that when exceeded during password verification will log a warning message.

## 10.2.0

### Minor Changes

- 803c518: chore: update core dependencies

### Patch Changes

- Updated dependencies [803c518]
  - @verdaccio/file-locking@10.2.0

## 10.1.0

### Minor Changes

- 4e9a3d0: feat: remove core-js from bundle

  By using babel.js core-js injects some requires that are not necessarily dependencies and fails on pnpm and yarn 2 due are strict. No need to add this feature so is removed.

  - https://babeljs.io/docs/en/babel-preset-env#usebuiltins

### Patch Changes

- Updated dependencies [4e9a3d0]
  - @verdaccio/file-locking@10.1.0

## 10.0.1

### Patch Changes

- 6134415: fix: update dependencies
- Updated dependencies [6134415]
  - @verdaccio/file-locking@10.0.1

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [10.0.0](https://github.com/verdaccio/monorepo/compare/v9.7.5...v10.0.0) (2021-03-29)

**Note:** Version bump only for package verdaccio-htpasswd

## [9.7.2](https://github.com/verdaccio/monorepo/compare/v9.7.1...v9.7.2) (2020-07-20)

**Note:** Version bump only for package verdaccio-htpasswd

## [9.7.1](https://github.com/verdaccio/monorepo/compare/v9.7.0...v9.7.1) (2020-07-10)

### Bug Fixes

- update dependencies ([#375](https://github.com/verdaccio/monorepo/issues/375)) ([1e7aeec](https://github.com/verdaccio/monorepo/commit/1e7aeec31b056979285e272793a95b8c75d57c77))

# [9.7.0](https://github.com/verdaccio/monorepo/compare/v9.6.1...v9.7.0) (2020-06-24)

**Note:** Version bump only for package verdaccio-htpasswd

## [9.6.1](https://github.com/verdaccio/monorepo/compare/v9.6.0...v9.6.1) (2020-06-07)

**Note:** Version bump only for package verdaccio-htpasswd

# [9.5.0](https://github.com/verdaccio/monorepo/compare/v9.4.1...v9.5.0) (2020-05-02)

**Note:** Version bump only for package verdaccio-htpasswd

## [9.4.1](https://github.com/verdaccio/monorepo/compare/v9.4.0...v9.4.1) (2020-04-30)

### Bug Fixes

- **verdaccio-htpasswd:** generate non-constant legacy 2 byte salt ([#357](https://github.com/verdaccio/monorepo/issues/357)) ([d522595](https://github.com/verdaccio/monorepo/commit/d522595122b7deaac8e3bc568f73658041811aaf))

# [9.4.0](https://github.com/verdaccio/monorepo/compare/v9.3.4...v9.4.0) (2020-03-21)

**Note:** Version bump only for package verdaccio-htpasswd

## [9.3.2](https://github.com/verdaccio/monorepo/compare/v9.3.1...v9.3.2) (2020-03-08)

### Bug Fixes

- update dependencies ([#332](https://github.com/verdaccio/monorepo/issues/332)) ([b6165ae](https://github.com/verdaccio/monorepo/commit/b6165aea9b7e4012477081eae68bfa7159c58f56))

## [9.3.1](https://github.com/verdaccio/monorepo/compare/v9.3.0...v9.3.1) (2020-02-23)

**Note:** Version bump only for package verdaccio-htpasswd

# [9.3.0](https://github.com/verdaccio/monorepo/compare/v9.2.0...v9.3.0) (2020-01-29)

**Note:** Version bump only for package verdaccio-htpasswd

# [9.0.0](https://github.com/verdaccio/monorepo/compare/v8.5.3...v9.0.0) (2020-01-07)

### chore

- update dependencies ([68add74](https://github.com/verdaccio/monorepo/commit/68add743159867f678ddb9168d2bc8391844de47))

### Features

- **eslint-config:** enable eslint curly ([#308](https://github.com/verdaccio/monorepo/issues/308)) ([91acb12](https://github.com/verdaccio/monorepo/commit/91acb121847018e737c21b367fcaab8baa918347))

### BREAKING CHANGES

- @verdaccio/eslint-config requires ESLint >=6.8.0 and Prettier >=1.19.1 to fix compatibility with overrides.extends config

## [8.5.2](https://github.com/verdaccio/monorepo/compare/v8.5.1...v8.5.2) (2019-12-25)

**Note:** Version bump only for package verdaccio-htpasswd

## [8.5.1](https://github.com/verdaccio/monorepo/compare/v8.5.0...v8.5.1) (2019-12-24)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.5.0](https://github.com/verdaccio/monorepo/compare/v8.4.2...v8.5.0) (2019-12-22)

**Note:** Version bump only for package verdaccio-htpasswd

## [8.4.2](https://github.com/verdaccio/monorepo/compare/v8.4.1...v8.4.2) (2019-11-23)

**Note:** Version bump only for package verdaccio-htpasswd

## [8.4.1](https://github.com/verdaccio/monorepo/compare/v8.4.0...v8.4.1) (2019-11-22)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.4.0](https://github.com/verdaccio/monorepo/compare/v8.3.0...v8.4.0) (2019-11-22)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.3.0](https://github.com/verdaccio/monorepo/compare/v8.2.0...v8.3.0) (2019-10-27)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.2.0](https://github.com/verdaccio/monorepo/compare/v8.2.0-next.0...v8.2.0) (2019-10-23)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.2.0-next.0](https://github.com/verdaccio/monorepo/compare/v8.1.4...v8.2.0-next.0) (2019-10-08)

### Bug Fixes

- fixed lint errors ([5e677f7](https://github.com/verdaccio/monorepo/commit/5e677f7))

## [8.1.2](https://github.com/verdaccio/monorepo/compare/v8.1.1...v8.1.2) (2019-09-29)

**Note:** Version bump only for package verdaccio-htpasswd

## [8.1.1](https://github.com/verdaccio/monorepo/compare/v8.1.0...v8.1.1) (2019-09-26)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.1.0](https://github.com/verdaccio/monorepo/compare/v8.0.1-next.1...v8.1.0) (2019-09-07)

**Note:** Version bump only for package verdaccio-htpasswd

## [8.0.1-next.1](https://github.com/verdaccio/monorepo/compare/v8.0.1-next.0...v8.0.1-next.1) (2019-08-29)

**Note:** Version bump only for package verdaccio-htpasswd

## [8.0.1-next.0](https://github.com/verdaccio/monorepo/compare/v8.0.0...v8.0.1-next.0) (2019-08-29)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.0.0](https://github.com/verdaccio/monorepo/compare/v8.0.0-next.4...v8.0.0) (2019-08-22)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.0.0-next.4](https://github.com/verdaccio/monorepo/compare/v8.0.0-next.3...v8.0.0-next.4) (2019-08-18)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.0.0-next.2](https://github.com/verdaccio/monorepo/compare/v8.0.0-next.1...v8.0.0-next.2) (2019-08-03)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.0.0-next.1](https://github.com/verdaccio/monorepo/compare/v8.0.0-next.0...v8.0.0-next.1) (2019-08-01)

**Note:** Version bump only for package verdaccio-htpasswd

# [8.0.0-next.0](https://github.com/verdaccio/monorepo/compare/v2.0.0...v8.0.0-next.0) (2019-08-01)

**Note:** Version bump only for package verdaccio-htpasswd

# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

# [2.0.0](https://github.com/verdaccio/verdaccio-htpasswd/compare/v2.0.0-beta.1...v2.0.0) (2019-04-14)

### Features

- drop node v6 suport ([d1d52e8](https://github.com/verdaccio/verdaccio-htpasswd/commit/d1d52e8))

<a name="2.0.0-beta.1"></a>

# [2.0.0-beta.1](https://github.com/verdaccio/verdaccio-htpasswd/compare/v2.0.0-beta.0...v2.0.0-beta.1) (2019-02-24)

### Bug Fixes

- package.json to reduce vulnerabilities ([259bdaf](https://github.com/verdaccio/verdaccio-htpasswd/commit/259bdaf))
- update [@verdaccio](https://github.com/verdaccio)/file-locking@1.0.0 ([ec0bbfd](https://github.com/verdaccio/verdaccio-htpasswd/commit/ec0bbfd))

<a name="2.0.0-beta.0"></a>

# [2.0.0-beta.0](https://github.com/verdaccio/verdaccio-htpasswd/compare/v1.0.1...v2.0.0-beta.0) (2019-02-03)

### Features

- migrate to typescript ([79f6937](https://github.com/verdaccio/verdaccio-htpasswd/commit/79f6937))
- remove Node6 from CircleCI ([d3a05ab](https://github.com/verdaccio/verdaccio-htpasswd/commit/d3a05ab))
- use verdaccio babel preset ([3a63f88](https://github.com/verdaccio/verdaccio-htpasswd/commit/3a63f88))

<a name="1.0.1"></a>

## [1.0.1](https://github.com/verdaccio/verdaccio-htpasswd/compare/v1.0.0...v1.0.1) (2018-09-30)

### Bug Fixes

- password hash & increase coverage ([6420c26](https://github.com/verdaccio/verdaccio-htpasswd/commit/6420c26))

<a name="1.0.0"></a>

# [1.0.0](https://github.com/verdaccio/verdaccio-htpasswd/compare/v0.2.2...v1.0.0) (2018-09-30)

### Bug Fixes

- adds error message for user registration ([0bab945](https://github.com/verdaccio/verdaccio-htpasswd/commit/0bab945))

### Features

- **change-passwd:** implement change password [#32](https://github.com/verdaccio/verdaccio-htpasswd/issues/32) ([830b143](https://github.com/verdaccio/verdaccio-htpasswd/commit/830b143))
