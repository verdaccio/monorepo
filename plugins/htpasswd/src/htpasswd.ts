import fs from 'fs';
import Path from 'path';
import buildDebug from 'debug';

import { Callback, AuthConf, Config, Logger, IPluginAuth, PluginOptions } from '@verdaccio/legacy-types';
import { unlockFile } from '@verdaccio/file-locking';

import {
  HtpasswdHashAlgorithm,
  HtpasswdHashConfig,
  verifyPassword,
  lockAndRead,
  parseHTPasswd,
  addUserToHTPasswd,
  changePasswordToHTPasswd,
  sanityCheck,
  DEFAULT_BCRYPT_ROUNDS,
} from './utils';

const debug = buildDebug('verdaccio:plugin:htpasswd');


export type HTPasswdConfig = {
  file: string;
  algorithm?: HtpasswdHashAlgorithm;
  rounds?: number;
  max_users?: number;
  slow_verify_ms?: number;
} & Config;

export const DEFAULT_SLOW_VERIFY_MS = 200;

/**
 * HTPasswd - Verdaccio auth class
 */
export default class HTPasswd implements IPluginAuth<HTPasswdConfig> {
  /**
   *
   * @param {*} config htpasswd file
   * @param {object} stuff config.yaml in object from
   */
  private users: {};
  private maxUsers: number;
  private path: string;
  private hashConfig: HtpasswdHashConfig;
  private slowVerifyMs: number;
  private logger: Logger;
  private lastTime: any;
  // constructor
  public constructor(config: AuthConf, options: PluginOptions<HTPasswdConfig>) {
    this.users = {};
    // verdaccio logger
    this.logger = options.logger;

    // all this "verdaccio_config" stuff is for b/w compatibility only
    this.maxUsers = config.max_users ? config.max_users : Infinity;

    let algorithm: HtpasswdHashAlgorithm;
    let rounds: number | undefined;

    if (config.algorithm === undefined) {
      // to avoid breaking changes we uses crypt, future version
      // of this plugin uses bcrypt by default
      // https://github.com/verdaccio/verdaccio/pull/2072#issuecomment-770235502
      algorithm = HtpasswdHashAlgorithm.crypt;
      this.logger.info(
        // eslint-disable-next-line max-len
        '"crypt" algorithm is deprecated consider switch to "bcrypt". Read more: https://github.com/verdaccio/monorepo/pull/580'
      );
    } else if (HtpasswdHashAlgorithm[config.algorithm] !== undefined) {
      algorithm = HtpasswdHashAlgorithm[config.algorithm];
    } else {
      throw new Error(`Invalid algorithm "${config.algorithm}"`);
    }

    this.lastTime = null;

    const { file } = config;
    debug('file: %s', file);
    if (!file) {
      throw new Error('should specify "file" in config');
    }

    debug(`password hash algorithm: ${algorithm}`);
    if (algorithm === HtpasswdHashAlgorithm.bcrypt) {
      rounds = config.rounds || DEFAULT_BCRYPT_ROUNDS;
    } else if (config.rounds !== undefined) {
      this.logger.warn({ algo: algorithm }, 'Option "rounds" is not valid for "@{algo}" algorithm');
    }

    this.hashConfig = {
      algorithm,
      rounds,
    };
    debug('config self_path: %s', options?.config?.self_path);
    debug('config configPath: %s', options?.config?.configPath);
    this.path = Path.resolve(Path.dirname(options.config.self_path), file);
    this.logger.info({ file: this.path }, 'using htpasswd file: @{file}');
    debug('htpasswd path:', this.path);
    this.slowVerifyMs = config.slow_verify_ms || DEFAULT_SLOW_VERIFY_MS;
  }

  /**
   * authenticate - Authenticate user.
   * @param {string} user
   * @param {string} password
   * @param {function} cb
   * @returns {function}
   */
  public authenticate(user: string, password: string, cb: Callback): void {
    this.reload(async (err) => {
      if (err) {
        return cb(err.code === 'ENOENT' ? null : err);
      }
      if (!this.users[user]) {
        return cb(null, false);
      }

      let passwordValid = false;
      try {
        const start = new Date();
        passwordValid = await verifyPassword(password, this.users[user]);
        const durationMs = new Date().getTime() - start.getTime();
        if (durationMs > this.slowVerifyMs) {
          this.logger.warn(
            { user, durationMs },
            'Password for user "@{user}" took @{durationMs}ms to verify'
          );
        }
      } catch ({ message }) {
        this.logger.error({ message }, 'Unable to verify user password: @{message}');
      }
      if (!passwordValid) {
        return cb(null, false);
      }

      // authentication succeeded!
      // return all usergroups this user has access to;
      // (this particular package has no concept of usergroups, so just return
      // user herself)
      return cb(null, [user]);
    });
  }

  /**
   * Add user
   * 1. lock file for writing (other processes can still read)
   * 2. reload .htpasswd
   * 3. write new data into .htpasswd.tmp
   * 4. move .htpasswd.tmp to .htpasswd
   * 5. reload .htpasswd
   * 6. unlock file
   *
   * @param {string} user
   * @param {string} password
   * @param {function} realCb
   * @returns {Promise<any>}
   */
  public adduser(user: string, password: string, realCb: Callback): void {
    const pathPass = this.path;
    sanityCheck(user, password, verifyPassword, this.users, this.maxUsers)
      .then((sanity) => {
        // preliminary checks, just to ensure that file won't be reloaded if it's
        // not needed
        if (sanity) {
          return realCb(sanity, false);
        }

        lockAndRead(pathPass, async (err, res): Promise<void> => {
          let locked = false;

          // callback that cleans up lock first
          const cb = (err): void => {
            if (locked) {
              unlockFile(pathPass, () => {
                // ignore any error from the unlock
                realCb(err, !err);
              });
            } else {
              realCb(err, !err);
            }
          };

          if (!err) {
            locked = true;
          }

          // ignore ENOENT errors, we'll just create .htpasswd in that case
          if (err && err.code !== 'ENOENT') {
            return cb(err);
          }
          const body = (res || '').toString('utf8');
          this.users = parseHTPasswd(body);

          // real checks, to prevent race conditions
          // parsing users after reading file.
          sanity = await sanityCheck(user, password, verifyPassword, this.users, this.maxUsers);

          if (sanity) {
            return cb(sanity);
          }

          try {
            this._writeFile(await addUserToHTPasswd(body, user, password, this.hashConfig), cb);
          } catch (err) {
            return cb(err);
          }
        });
      })
      .catch((err) => realCb(err));
  }

  /**
   * Reload users
   * @param {function} callback
   */
  public reload(callback: Callback): void {
    fs.stat(this.path, (err, stats) => {
      if (err) {
        return callback(err);
      }
      if (this.lastTime === stats.mtime) {
        return callback();
      }

      this.lastTime = stats.mtime;

      fs.readFile(this.path, 'utf8', (err, buffer) => {
        if (err) {
          return callback(err);
        }

        Object.assign(this.users, parseHTPasswd(buffer));
        callback();
      });
    });
  }

  private _stringToUt8(authentication: string): string {
    return (authentication || '').toString();
  }

  private _writeFile(body: string, cb: Callback): void {
    fs.writeFile(this.path, body, (err) => {
      if (err) {
        cb(err);
      } else {
        this.reload(() => {
          cb(null);
        });
      }
    });
  }

  /**
   * changePassword - change password for existing user.
   * @param {string} user
   * @param {string} password
   * @param {string} newPassword
   * @param {function} realCb
   * @returns {function}
   */
  public changePassword(
    user: string,
    password: string,
    newPassword: string,
    realCb: Callback
  ): void {
    lockAndRead(this.path, (err, res) => {
      let locked = false;
      const pathPassFile = this.path;

      // callback that cleans up lock first
      const cb = (err): void => {
        if (locked) {
          unlockFile(pathPassFile, () => {
            // ignore any error from the unlock
            realCb(err, !err);
          });
        } else {
          realCb(err, !err);
        }
      };

      if (!err) {
        locked = true;
      }

      if (err && err.code !== 'ENOENT') {
        return cb(err);
      }

      const body = this._stringToUt8(res);
      this.users = parseHTPasswd(body);

      try {
        changePasswordToHTPasswd(body, user, password, newPassword, this.hashConfig)
          .then((content) => {
            this._writeFile(content, cb);
          })
          .catch((err) => {
            cb(err);
          });
      } catch (err) {
        return cb(err);
      }
    });
  }
}
