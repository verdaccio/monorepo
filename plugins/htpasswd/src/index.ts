import HTPasswd, { HTPasswdConfig } from './htpasswd';

/**
 * A new instance of HTPasswd class.
 * @param {object} config
 * @param {object} stuff
 * @returns {object}
 */
export default function (config, stuff): HTPasswd {
  return new HTPasswd(config, stuff);
}

export { HTPasswd, HTPasswdConfig };
