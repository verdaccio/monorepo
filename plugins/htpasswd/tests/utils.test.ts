// @ts-ignore: Module has no default export
import crypto from 'crypto';
import MockDate from 'mockdate';

import { DEFAULT_BCRYPT_ROUNDS } from '../src/htpasswd';
import {
  HtpasswdHashAlgorithm,
  addUserToHTPasswd,
  changePasswordToHTPasswd,
  generateHtpasswdLine,
  lockAndRead,
  parseHTPasswd,
  sanityCheck,
  verifyPassword,
} from '../src/utils';

const mockReadFile = jest.fn();
const mockUnlockFile = jest.fn();

const defaultHashConfig = {
  algorithm: HtpasswdHashAlgorithm.bcrypt,
  rounds: DEFAULT_BCRYPT_ROUNDS,
};

const mockTimeAndRandomBytes = () => {
  MockDate.set('2018-01-14T11:17:40.712Z');
  crypto.randomBytes = jest.fn(() => {
    return {
      toString: (): string => '$6',
    };
  });
  Math.random = jest.fn(() => 0.38849);
};

jest.mock('@verdaccio/file-locking', () => ({
  readFile: () => mockReadFile(),
  unlockFile: () => mockUnlockFile(),
}));

describe('parseHTPasswd', () => {
  it('should parse the password for a single line', () => {
    const input = 'test:$6b9MlB3WUELU:autocreated 2017-11-06T18:17:21.957Z';
    const output = { test: '$6b9MlB3WUELU' };
    expect(parseHTPasswd(input)).toEqual(output);
  });

  it('should parse the password for two lines', () => {
    const input = `user1:$6b9MlB3WUELU:autocreated 2017-11-06T18:17:21.957Z
user2:$6FrCaT/v0dwE:autocreated 2017-12-14T13:30:20.838Z`;
    const output = { user1: '$6b9MlB3WUELU', user2: '$6FrCaT/v0dwE' };
    expect(parseHTPasswd(input)).toEqual(output);
  });

  it('should parse the password for multiple lines', () => {
    const input = `user1:$6b9MlB3WUELU:autocreated 2017-11-06T18:17:21.957Z
user2:$6FrCaT/v0dwE:autocreated 2017-12-14T13:30:20.838Z
user3:$6FrCdfd\v0dwE:autocreated 2017-12-14T13:30:20.838Z
user4:$6FrCasdvppdwE:autocreated 2017-12-14T13:30:20.838Z`;
    const output = {
      user1: '$6b9MlB3WUELU',
      user2: '$6FrCaT/v0dwE',
      user3: '$6FrCdfd\v0dwE',
      user4: '$6FrCasdvppdwE',
    };
    expect(parseHTPasswd(input)).toEqual(output);
  });
});

describe('verifyPassword', () => {
  it('should verify the MD5/Crypt3 password with true', async () => {
    const input = ['test', '$apr1$sKXK9.lG$rZ4Iy63Vtn8jF9/USc4BV0'];
    expect(await verifyPassword(input[0], input[1])).toBeTruthy();
  });
  it('should verify the MD5/Crypt3 password with false', async () => {
    const input = ['testpasswordchanged', '$apr1$sKXK9.lG$rZ4Iy63Vtn8jF9/USc4BV0'];
    expect(await verifyPassword(input[0], input[1])).toBeFalsy();
  });
  it('should verify the plain password with true', async () => {
    const input = ['testpasswordchanged', '{PLAIN}testpasswordchanged'];
    expect(await verifyPassword(input[0], input[1])).toBeTruthy();
  });
  it('should verify the plain password with false', async () => {
    const input = ['testpassword', '{PLAIN}testpasswordchanged'];
    expect(await verifyPassword(input[0], input[1])).toBeFalsy();
  });
  it('should verify the crypto SHA password with true', async () => {
    const input = ['testpassword', '{SHA}i7YRj4/Wk1rQh2o740pxfTJwj/0='];
    expect(await verifyPassword(input[0], input[1])).toBeTruthy();
  });
  it('should verify the crypto SHA password with false', async () => {
    const input = ['testpasswordchanged', '{SHA}i7YRj4/Wk1rQh2o740pxfTJwj/0='];
    expect(await verifyPassword(input[0], input[1])).toBeFalsy();
  });
  it('should verify the bcrypt password with true', async () => {
    const input = ['testpassword', '$2y$04$Wqed4yN0OktGbiUdxSTwtOva1xfESfkNIZfcS9/vmHLsn3.lkFxJO'];
    expect(await verifyPassword(input[0], input[1])).toBeTruthy();
  });
  it('should verify the bcrypt password with false', async () => {
    const input = [
      'testpasswordchanged',
      '$2y$04$Wqed4yN0OktGbiUdxSTwtOva1xfESfkNIZfcS9/vmHLsn3.lkFxJO',
    ];
    expect(await verifyPassword(input[0], input[1])).toBeFalsy();
  });
});

describe('generateHtpasswdLine', () => {
  beforeAll(mockTimeAndRandomBytes);

  const [user, passwd] = ['username', 'password'];

  it('should correctly generate line for md5', () => {
    const md5Conf = { algorithm: HtpasswdHashAlgorithm.md5 };
    expect(generateHtpasswdLine(user, passwd, md5Conf)).toMatchSnapshot();
  });

  it('should correctly generate line for sha1', () => {
    const sha1Conf = { algorithm: HtpasswdHashAlgorithm.sha1 };
    expect(generateHtpasswdLine(user, passwd, sha1Conf)).toMatchSnapshot();
  });

  it('should correctly generate line for crypt', () => {
    const cryptConf = { algorithm: HtpasswdHashAlgorithm.crypt };
    expect(generateHtpasswdLine(user, passwd, cryptConf)).toMatchSnapshot();
  });

  it('should correctly generate line for bcrypt', () => {
    const bcryptAlgoConfig = {
      algorithm: HtpasswdHashAlgorithm.bcrypt,
      rounds: 2,
    };
    expect(generateHtpasswdLine(user, passwd, bcryptAlgoConfig)).toMatchSnapshot();
  });
});

describe('addUserToHTPasswd - bcrypt', () => {
  beforeAll(mockTimeAndRandomBytes);

  it('should add new htpasswd to the end', () => {
    const input = ['', 'username', 'password'];
    expect(addUserToHTPasswd(input[0], input[1], input[2], defaultHashConfig)).toMatchSnapshot();
  });

  it('should add new htpasswd to the end in multiline input', () => {
    const body = `test1:$6b9MlB3WUELU:autocreated 2017-11-06T18:17:21.957Z
    test2:$6FrCaT/v0dwE:autocreated 2017-12-14T13:30:20.838Z`;
    const input = [body, 'username', 'password'];
    expect(addUserToHTPasswd(input[0], input[1], input[2], defaultHashConfig)).toMatchSnapshot();
  });

  it('should throw an error for incorrect username with space', () => {
    const [a, b, c] = ['', 'firstname lastname', 'password'];
    expect(() => addUserToHTPasswd(a, b, c, defaultHashConfig)).toThrowErrorMatchingSnapshot();
  });
});

describe('lockAndRead', () => {
  it('should call the readFile method', () => {
    // console.log(fileLocking);
    // const spy = jest.spyOn(fileLocking, 'readFile');
    const cb = (): void => {};
    lockAndRead('.htpasswd', cb);
    expect(mockReadFile).toHaveBeenCalled();
  });
});

describe('sanityCheck', () => {
  let users;

  beforeEach(() => {
    users = { test: '$6FrCaT/v0dwE' };
  });

  test('should throw error for user already exists', async () => {
    const verifyFn = jest.fn();
    const input = await sanityCheck('test', users.test, verifyFn, users, Infinity);
    expect(input.status).toEqual(401);
    expect(input.message).toEqual('unauthorized access');
    expect(verifyFn).toHaveBeenCalled();
  });

  test('should throw error for registration disabled of users', async () => {
    const verifyFn = (): void => {};
    const input = await sanityCheck('username', users.test, verifyFn, users, -1);
    expect(input.status).toEqual(409);
    expect(input.message).toEqual('user registration disabled');
  });

  test('should throw error max number of users', async () => {
    const verifyFn = (): void => {};
    const input = await sanityCheck('username', users.test, verifyFn, users, 1);
    expect(input.status).toEqual(403);
    expect(input.message).toEqual('maximum amount of users reached');
  });

  test('should not throw anything and sanity check', async () => {
    const verifyFn = (): void => {};
    const input = await sanityCheck('username', users.test, verifyFn, users, 2);
    expect(input).toBeNull();
  });

  test('should throw error for required username field', async () => {
    const verifyFn = (): void => {};
    const input = await sanityCheck(undefined, users.test, verifyFn, users, 2);
    expect(input.message).toEqual('username and password is required');
    expect(input.status).toEqual(400);
  });

  test('should throw error for required password field', async () => {
    const verifyFn = (): void => {};
    const input = await sanityCheck('username', undefined, verifyFn, users, 2);
    expect(input.message).toEqual('username and password is required');
    expect(input.status).toEqual(400);
  });

  test('should throw error for required username & password fields', async () => {
    const verifyFn = (): void => {};
    const input = await sanityCheck(undefined, undefined, verifyFn, users, 2);
    expect(input.message).toEqual('username and password is required');
    expect(input.status).toEqual(400);
  });

  test('should throw error for existing username and password', async () => {
    const verifyFn = jest.fn(() => true);
    const input = await sanityCheck('test', users.test, verifyFn, users, 2);
    expect(input.status).toEqual(409);
    expect(input.message).toEqual('username is already registered');
    expect(verifyFn).toHaveBeenCalledTimes(1);
  });

  test(
    'should throw error for existing username and password with max number ' + 'of users reached',
    async () => {
      const verifyFn = jest.fn(() => true);
      const input = await sanityCheck('test', users.test, verifyFn, users, 1);
      expect(input.status).toEqual(409);
      expect(input.message).toEqual('username is already registered');
      expect(verifyFn).toHaveBeenCalledTimes(1);
    }
  );
});

describe('changePasswordToHTPasswd', () => {
  test('should throw error for wrong password', async () => {
    const body = 'test:$6b9MlB3WUELU:autocreated 2017-11-06T18:17:21.957Z';

    try {
      await changePasswordToHTPasswd(
        body,
        'test',
        'somerandompassword',
        'newPassword',
        defaultHashConfig
      );
    } catch (error: any) {
      expect(error.message).toEqual(
        `Unable to change password for user 'test': invalid old password`
      );
    }
  });

  test('should throw error when user does not exist', async () => {
    const body = 'test:$6b9MlB3WUELU:autocreated 2017-11-06T18:17:21.957Z';

    try {
      await changePasswordToHTPasswd(
        body,
        'test2',
        'somerandompassword',
        'newPassword',
        defaultHashConfig
      );
    } catch (error: any) {
      expect(error.message).toEqual(
        `Unable to change password for user 'test2': user does not currently exist`
      );
    }
  });

  test('should change the password', async () => {
    const body = 'root:$6qLTHoPfGLy2:autocreated 2018-08-20T13:38:12.164Z';

    expect(
      await changePasswordToHTPasswd(body, 'root', 'demo123', 'newPassword', defaultHashConfig)
    ).toMatchSnapshot();
  });
});
