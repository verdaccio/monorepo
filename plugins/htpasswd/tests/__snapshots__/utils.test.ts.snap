// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUserToHTPasswd - bcrypt should add new htpasswd to the end 1`] = `
"username:$2a$10$......................7zqaLmaKtn.i7IjPfuPGY2Ah/mNM6Sy:autocreated 2018-01-14T11:17:40.712Z
"
`;

exports[`addUserToHTPasswd - bcrypt should add new htpasswd to the end in multiline input 1`] = `
"test1:$6b9MlB3WUELU:autocreated 2017-11-06T18:17:21.957Z
    test2:$6FrCaT/v0dwE:autocreated 2017-12-14T13:30:20.838Z
username:$2a$10$......................7zqaLmaKtn.i7IjPfuPGY2Ah/mNM6Sy:autocreated 2018-01-14T11:17:40.712Z
"
`;

exports[`addUserToHTPasswd - bcrypt should throw an error for incorrect username with space 1`] = `"username should not contain non-uri-safe characters"`;

exports[`changePasswordToHTPasswd should change the password 1`] = `
"root:$2a$10$......................0qqDmeqkAfPx68M2ArX8hVzcVNft5Ha:autocreated 2018-01-14T11:17:40.712Z
"
`;

exports[`generateHtpasswdLine should correctly generate line for bcrypt 1`] = `
"username:$2a$04$......................LAtw7/ohmmBAhnXqmkuIz83Rl5Qdjhm:autocreated 2018-01-14T11:17:40.712Z
"
`;

exports[`generateHtpasswdLine should correctly generate line for crypt 1`] = `
"username:$66to3JK5RgZM:autocreated 2018-01-14T11:17:40.712Z
"
`;

exports[`generateHtpasswdLine should correctly generate line for md5 1`] = `
"username:$apr1$MMMMMMMM$2lGUwLC3NFfN74jH51z1W.:autocreated 2018-01-14T11:17:40.712Z
"
`;

exports[`generateHtpasswdLine should correctly generate line for sha1 1`] = `
"username:{SHA}W6ph5Mm5Pz8GgiULbPgzG37mj9g=:autocreated 2018-01-14T11:17:40.712Z
"
`;
